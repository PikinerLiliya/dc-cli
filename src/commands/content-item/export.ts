import { Arguments, Argv } from 'yargs';
import { ConfigurationParameters } from '../configure';
import dynamicContentClientFactory from '../../services/dynamic-content-client-factory';
import { FileLog } from '../../common/file-log';
import { dirname, join } from 'path';
import { equalsOrRegex } from '../../common/filter/filter';
import sanitize from 'sanitize-filename';
import { uniqueFilenamePath, writeJsonToFile } from '../../services/export.service';

import { mkdir, writeFile, exists, lstat } from 'fs';
import { promisify } from 'util';
import { ExportItemBuilderOptions } from '../../interfaces/export-item-builder-options.interface';
import paginator from '../../common/dc-management-sdk-js/paginator';
import { ContentItem, Folder, DynamicContent, Hub, ContentRepository } from 'dc-management-sdk-js';

import { ensureDirectoryExists } from '../../common/import/directory-utils';
import { ContentDependancyTree } from '../../common/content-item/content-dependancy-tree';
import { ContentMapping } from '../../common/content-item/content-mapping';
import { getDefaultLogPath } from '../../common/log-helpers';
import { AmplienceSchemaValidator } from '../../common/content-item/amplience-schema-validator';

export const command = 'export <dir>';

export const desc = 'Export Content Items';

export const LOG_FILENAME = (platform: string = process.platform): string =>
  getDefaultLogPath('item', 'export', platform);

export const builder = (yargs: Argv): void => {
  yargs
    .positional('dir', {
      describe: 'Output directory for the exported Content Items',
      type: 'string',
      requiresArg: true
    })
    .option('repoId', {
      type: 'string',
      describe:
        'Export content from within a given repository. Directory structure will start at the specified repository. Will automatically export all contained folders.'
    })
    .option('folderId', {
      type: 'string',
      describe:
        'Export content from within a given folder. Directory structure will start at the specified folder. Can be used in addition to repoId.'
    })
    .option('schemaId', {
      type: 'string',
      describe:
        'Export content with a given or matching Schema ID. A regex can be provided, surrounded with forward slashes. Can be used in combination with other filters.'
    })
    .option('name', {
      type: 'string',
      describe:
        'Export content with a given or matching Name. A regex can be provided, surrounded with forward slashes. Can be used in combination with other filters.'
    })
    .option('logFile', {
      type: 'string',
      default: LOG_FILENAME,
      describe: 'Path to a log file to write to.'
    });
};

const getOrAddFolderPath = async (
  folderToPathMap: Map<string, string>,
  client: DynamicContent,
  folderOrId: Folder | string | undefined,
  log: FileLog,
  baseDir?: string
): Promise<string> => {
  if (folderOrId == null) return '';
  const id = typeof folderOrId === 'string' ? folderOrId : (folderOrId.id as string);

  const mapResult = folderToPathMap.get(id);
  if (mapResult !== undefined) {
    return mapResult;
  }

  // Build the path for this folder.
  const folder = typeof folderOrId === 'string' ? await client.folders.get(folderOrId) : folderOrId;

  const name = sanitize(folder.name as string);
  let path: string;
  try {
    const parent = await folder.related.folders.parent();

    path = `${join(await getOrAddFolderPath(folderToPathMap, client, parent, log), name)}`;
  } catch {
    log.appendLine(`Could not determine path for ${folder.name}. Placing in base directory.`);
    path = `${name}`;
  }

  if (baseDir != null) {
    path = join(baseDir, path);
  }

  folderToPathMap.set(id, path);
  return path;
};

const getContentItems = async (
  folderToPathMap: Map<string, string>,
  client: DynamicContent,
  hub: Hub,
  dir: string,
  log: FileLog,
  repoId?: string | string[],
  folderId?: string | string[]
): Promise<{ path: string; item: ContentItem }[]> => {
  const items: { path: string; item: ContentItem }[] = [];

  const folderIds = typeof folderId === 'string' ? [folderId] : folderId || [];

  const repoItems: ContentItem[] = [];

  const repoIds = typeof repoId === 'string' ? [repoId] : repoId || [];

  const repositories = await (repoId != null || folderId != null
    ? Promise.all(repoIds.map(id => client.contentRepositories.get(id)))
    : paginator(hub.related.contentRepositories.list));

  let specifyBasePaths = repositories.length + folderIds.length > 1;

  for (let i = 0; i < repositories.length; i++) {
    const repository = repositories[i];
    const baseDir = specifyBasePaths ? `${sanitize(repository.label as string)}/` : '';
    await ensureDirectoryExists(join(dir, baseDir));
    const newFolders = await paginator(repository.related.folders.list);
    newFolders.forEach(folder => {
      if (folderIds.indexOf(folder.id as string) === -1) {
        folderIds.push(folder.id as string);
      }
      folderToPathMap.set(folder.id as string, join(baseDir, `${sanitize(folder.name as string)}/`));
    });

    // Add content items in repo base folder. Cache the other items so we don't have to request them again.
    let newItems: ContentItem[];
    try {
      const allItems = await paginator(repository.related.contentItems.list, { status: 'ACTIVE' });
      Array.prototype.push.apply(repoItems, allItems);
      newItems = allItems.filter(item => item.folderId == null);
    } catch (e) {
      console.error(`Error getting items from repository ${repository.name} (${repository.id}): ${e.toString()}`);
      continue;
    }

    Array.prototype.push.apply(items, newItems.map(item => ({ item, path: baseDir })));
  }

  const parallelism = 10;
  const folders = await Promise.all(folderIds.map(id => client.folders.get(id)));
  log.appendLine(`Found ${folders.length} base folders.`);

  specifyBasePaths = specifyBasePaths || folders.length > 1;

  const nextFolders: Folder[] = [];
  let processFolders = folders;
  let baseFolder = true;

  while (processFolders.length > 0) {
    const promises = processFolders.map(
      async (folder: Folder): Promise<void> => {
        if (baseFolder) {
          if (!folderToPathMap.has(folder.id as string)) {
            folderToPathMap.set(folder.id as string, specifyBasePaths ? `${sanitize(folder.name as string)}/` : '');
          }
        }
        const path = await getOrAddFolderPath(folderToPathMap, client, folder, log);
        log.appendLine(`Processing ${path}...`);

        let newItems: ContentItem[];
        // If we already have seen items in this folder, use those. Otherwise try get them explicitly.
        // This may happen for folders in selected repositories if they are empty, but it will be a no-op (and is unavoidable).
        newItems = repoItems.filter(item => item.folderId == folder.id);
        if (newItems.length == 0) {
          log.appendLine(`Fetching additional folder: ${folder.name}`);
          try {
            newItems = (await paginator(folder.related.contentItems.list)).filter(item => item.status === 'ACTIVE');
          } catch (e) {
            console.error(`Error getting items from folder ${folder.name} (${folder.id}): ${e.toString()}`);
            return;
          }
        }
        Array.prototype.push.apply(items, newItems.map(item => ({ item, path: path })));

        try {
          const subfolders = await paginator(folder.related.folders.list);
          Array.prototype.push.apply(nextFolders, subfolders);
        } catch (e) {
          console.error(`Error getting subfolders from folder ${folder.name} (${folder.id}): ${e.toString()}`);
        }
      }
    );

    await Promise.all(promises);

    baseFolder = false;
    processFolders = nextFolders.splice(0, Math.min(nextFolders.length, parallelism));
  }
  return items;
};

export const handler = async (argv: Arguments<ExportItemBuilderOptions & ConfigurationParameters>): Promise<void> => {
  const { dir, repoId, folderId, schemaId, name, logFile } = argv;

  const dummyRepo = new ContentRepository();

  const folderToPathMap: Map<string, string> = new Map();
  const client = dynamicContentClientFactory(argv);
  const log = typeof logFile === 'string' || logFile == null ? new FileLog(logFile) : logFile;
  const hub = await client.hubs.get(argv.hubId);

  log.appendLine('Retrieving content items, please wait.');
  let items = await getContentItems(folderToPathMap, client, hub, dir, log, repoId, folderId);

  // Filter using the schemaId and name, if present.
  if (schemaId != null) {
    const schemaIds: string[] = Array.isArray(schemaId) ? schemaId : [schemaId];
    items = items.filter(
      ({ item }: { item: ContentItem }) => schemaIds.findIndex(id => equalsOrRegex(item.body._meta.schema, id)) !== -1
    );
  }
  if (name != null) {
    const names: string[] = Array.isArray(name) ? name : [name];
    items = items.filter(
      ({ item }: { item: ContentItem }) => names.findIndex(name => equalsOrRegex(item.label as string, name)) !== -1
    );
  }

  log.appendLine('Scanning for dependancies.');
  const tree = new ContentDependancyTree(
    items.map(item => ({ repo: dummyRepo, content: item.item })),
    new ContentMapping()
  );

  const missingIDs = new Set<string>();
  const invalidContentItems = tree.filterAny(item => {
    const missingDeps = item.dependancies.filter(dep => !tree.byId.has(dep.dependancy.id as string));
    missingDeps.forEach(dep => {
      if (dep.dependancy.id != null) {
        missingIDs.add(dep.dependancy.id);
      }
    });
    return missingDeps.length > 0;
  });

  if (invalidContentItems) {
    // There are missing content items. We'll need to fetch them and see what their deal is.
    const missingIdArray = Array.from(missingIDs);
    for (let i = 0; i < missingIdArray.length; i++) {
      let item: ContentItem | null = null;

      try {
        item = await client.contentItems.get(missingIdArray[i]);
      } catch {}

      if (item != null) {
        if (item.status === 'ACTIVE') {
          // The item is active and should probably be included.
          const path = '_dependancies/';
          items.push({ item, path });

          log.appendLine(`Referenced content '${item.label}' added to the export.`);
        } else {
          // The item is archived and should not be included. Make a note to the user.
          log.appendLine(`Referenced content '${item.label}' is archived, so was not exported.`);
        }
      } else {
        log.appendLine(`Referenced content ${missingIdArray[i]} does not exist.`);
      }
    }
  }

  log.appendLine('Saving content items.');
  const filenames: string[] = [];

  const schemas = await paginator(hub.related.contentTypeSchema.list);

  const validator = new AmplienceSchemaValidator(schemas);

  for (let i = 0; i < items.length; i++) {
    const { item, path } = items[i];

    try {
      const errors = await validator.validate(item.body);
      if (errors.length > 0) {
        log.appendLine(
          `WARNING: ${item.label} does not validate under the available schema. It may not import correctly.`
        );
        log.appendLine(JSON.stringify(errors, null, 2));
      }
    } catch (e) {
      log.appendLine(`WARNING: Could not validate ${item.label} as there is a problem with the schema: ${e}`);
    }

    let resolvedPath: string;
    resolvedPath = path;

    const directory = join(dir, resolvedPath);
    resolvedPath = uniqueFilenamePath(directory, `${sanitize(item.label as string)}`, 'json', filenames);
    filenames.push(resolvedPath);
    log.appendLine(resolvedPath);
    await ensureDirectoryExists(directory);

    writeJsonToFile(resolvedPath, item);
  }

  log.close();
};
