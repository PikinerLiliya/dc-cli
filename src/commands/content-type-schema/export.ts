import { Arguments, Argv } from 'yargs';
import { ConfigurationParameters } from '../configure';
import dynamicContentClientFactory from '../../services/dynamic-content-client-factory';
import paginator from '../../common/dc-management-sdk-js/paginator';
import { ContentTypeSchema } from 'dc-management-sdk-js';
import { createStream } from 'table';
import { streamTableOptions } from '../../common/table/table.consts';
import { TableStream } from '../../interfaces/table.interface';
import chalk from 'chalk';
import {
  ExportResult,
  nothingExportedExit,
  promptToOverwriteExports,
  uniqueFilename,
  writeJsonToFile
} from '../../services/export.service';
import { loadJsonFromDirectory } from '../../services/import.service';
import { ExportBuilderOptions } from '../../interfaces/export-builder-options.interface';
import * as path from 'path';
import * as fs from 'fs';
import { resolveSchemaBody } from '../../services/resolve-schema-body';

export const command = 'export <dir>';

export const desc = 'Export Content Type Schemas';

export const builder = (yargs: Argv): void => {
  yargs
    .positional('dir', {
      describe: 'Output directory for the exported Content Type Schema definitions',
      type: 'string'
    })
    .option('schemaId', {
      type: 'string',
      describe:
        'The Schema ID of a Content Type Schema to be exported.\nIf no --schemaId option is given, all content type schemas for the hub are exported.\nA single --schemaId option may be given to export a single content type schema.\nMultiple --schemaId options may be given to export multiple content type schemas at the same time.',
      requiresArg: true
    });
};

const equals = (a: ContentTypeSchema, b: ContentTypeSchema): boolean =>
  a.schemaId === b.schemaId && a.body === b.body && a.validationLevel === b.validationLevel;

interface ExportRecord {
  readonly filename: string;
  readonly status: ExportResult;
  readonly contentTypeSchema: ContentTypeSchema;
}

type ExportsMap = {
  schemaId: string;
  filename: string;
};

const SCHEMA_DIR = 'schemas';

export const generateSchemaPath = (filepath: string): string =>
  SCHEMA_DIR + path.sep + path.basename(filepath).replace('.json', '-schema.json');

export const writeSchemaBody = (filename: string, body?: string): void => {
  if (!body) {
    return;
  }

  const dir = path.dirname(filename);
  if (fs.existsSync(dir)) {
    const dirStat = fs.lstatSync(dir);
    if (!dirStat || !dirStat.isDirectory()) {
      throw new Error(`Unable to write schema to "${filename}" as "${dir}" is not a directory.`);
    }
  } else {
    fs.mkdirSync(dir);
  }

  try {
    fs.writeFileSync(filename, body);
  } catch {
    throw new Error(`Unable to write file: ${filename}, aborting export`);
  }
};

export const getExportRecordForContentTypeSchema = (
  contentTypeSchema: ContentTypeSchema,
  outputDir: string,
  previouslyExportedContentTypeSchemas: { [filename: string]: ContentTypeSchema }
): ExportRecord => {
  const indexOfExportedContentTypeSchema = Object.values(previouslyExportedContentTypeSchemas).findIndex(
    c => c.schemaId === contentTypeSchema.schemaId
  );
  if (indexOfExportedContentTypeSchema < 0) {
    return {
      filename: uniqueFilename(
        outputDir,
        contentTypeSchema.schemaId,
        'json',
        Object.keys(previouslyExportedContentTypeSchemas)
      ),
      status: 'CREATED',
      contentTypeSchema
    };
  }
  const filename = Object.keys(previouslyExportedContentTypeSchemas)[indexOfExportedContentTypeSchema];
  const previouslyExportedContentTypeSchema = Object.values(previouslyExportedContentTypeSchemas)[
    indexOfExportedContentTypeSchema
  ];

  if (equals(previouslyExportedContentTypeSchema, contentTypeSchema)) {
    return { filename, status: 'UP-TO-DATE', contentTypeSchema };
  }
  return {
    filename,
    status: 'UPDATED',
    contentTypeSchema
  };
};

export const filterContentTypeSchemasBySchemaId = (
  listToFilter: ContentTypeSchema[],
  listToMatch: string[] = []
): ContentTypeSchema[] => {
  if (listToMatch.length === 0) {
    return listToFilter;
  }

  const unmatchedIdList: string[] = listToMatch.filter(id => !listToFilter.some(schema => schema.schemaId === id));
  if (unmatchedIdList.length > 0) {
    throw new Error(
      `The following schema ID(s) could not be found: [${unmatchedIdList
        .map(u => `'${u}'`)
        .join(', ')}].\nNothing was exported, exiting.`
    );
  }

  return listToFilter.filter(schema => listToMatch.some(id => schema.schemaId === id));
};

export const getContentTypeSchemaExports = (
  outputDir: string,
  previouslyExportedContentTypeSchemas: { [filename: string]: ContentTypeSchema },
  contentTypeSchemasBeingExported: ContentTypeSchema[]
): [ExportRecord[], ExportsMap[]] => {
  const allExports: ExportRecord[] = [];
  const updatedExportsMap: ExportsMap[] = []; // uri x filename
  for (const contentTypeSchema of contentTypeSchemasBeingExported) {
    if (!contentTypeSchema.schemaId) {
      continue;
    }

    const exportRecord = getExportRecordForContentTypeSchema(
      contentTypeSchema,
      outputDir,
      previouslyExportedContentTypeSchemas
    );
    allExports.push(exportRecord);
    if (exportRecord.status === 'UPDATED') {
      updatedExportsMap.push({ schemaId: contentTypeSchema.schemaId, filename: exportRecord.filename });
    }
  }
  return [allExports, updatedExportsMap];
};

export const processContentTypeSchemas = async (
  outputDir: string,
  previouslyExportedContentTypeSchemas: { [filename: string]: ContentTypeSchema },
  storedContentTypeSchemas: ContentTypeSchema[]
): Promise<void> => {
  if (storedContentTypeSchemas.length === 0) {
    nothingExportedExit('No content type schemas to export from this hub, exiting.\n');
  }

  const [allExports, updatedExportsMap] = getContentTypeSchemaExports(
    outputDir,
    previouslyExportedContentTypeSchemas,
    storedContentTypeSchemas
  );
  if (
    allExports.length === 0 ||
    (Object.keys(updatedExportsMap).length > 0 && !(await promptToOverwriteExports(updatedExportsMap)))
  ) {
    nothingExportedExit();
  }

  const tableStream = (createStream(streamTableOptions) as unknown) as TableStream;
  tableStream.write([chalk.bold('file'), chalk.bold('schemaId'), chalk.bold('result')]);
  for (const { filename, status, contentTypeSchema } of allExports) {
    if (status !== 'UP-TO-DATE') {
      delete contentTypeSchema.id; // do not export id
      const schemaBody = contentTypeSchema.body;
      const schemaBodyFilename = generateSchemaPath(filename);
      contentTypeSchema.body = './' + schemaBodyFilename;
      writeSchemaBody(outputDir + path.sep + schemaBodyFilename, schemaBody);
      writeJsonToFile(filename, new ContentTypeSchema(contentTypeSchema));
    }
    tableStream.write([filename, contentTypeSchema.schemaId || '', status]);
  }
  process.stdout.write('\n');
};

export const handler = async (argv: Arguments<ExportBuilderOptions & ConfigurationParameters>): Promise<void> => {
  const { dir, schemaId } = argv;
  const [contentTypeSchemas] = await resolveSchemaBody(
    loadJsonFromDirectory<ContentTypeSchema>(dir, ContentTypeSchema),
    dir
  );
  const client = dynamicContentClientFactory(argv);
  const hub = await client.hubs.get(argv.hubId);
  const storedContentTypeSchemas = await paginator(hub.related.contentTypeSchema.list);
  const schemaIdArray: string[] = schemaId ? (Array.isArray(schemaId) ? schemaId : [schemaId]) : [];
  const filteredContentTypeSchemas = filterContentTypeSchemasBySchemaId(storedContentTypeSchemas, schemaIdArray);
  await processContentTypeSchemas(dir, contentTypeSchemas, filteredContentTypeSchemas);
};
