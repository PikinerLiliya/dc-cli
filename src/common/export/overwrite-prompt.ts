import DataPresenter from '../../view/data-presenter';
import readline from 'readline';

export const promptToOverwriteExports = (updatedExportsMap: { [key: string]: string }[]): Promise<boolean> => {
  return new Promise((resolve): void => {
    process.stdout.write('The following files will be overwritten:');
    // display updatedExportsMap as a table of uri x filename
    new DataPresenter(updatedExportsMap).render();

    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });

    rl.question('Do you want to continue (y/n)?: ', answer => {
      rl.close();
      return resolve(answer === 'y');
    });
  });
};
