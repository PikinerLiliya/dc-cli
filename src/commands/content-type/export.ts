import { Arguments, Argv } from 'yargs';
import { ConfigurationParameters } from '../configure';
import dynamicContentClientFactory from '../../services/dynamic-content-client-factory';
import paginator from '../../common/dc-management-sdk-js/paginator';
import { ContentType } from 'dc-management-sdk-js';
import { createStream } from 'table';
import { streamTableOptions } from '../../common/table/table.consts';
import { TableStream } from '../../interfaces/table.interface';
import chalk from 'chalk';
import { ExportResult, uniqueFilename, writeJsonToFile } from '../../services/export.service';
import { loadJsonFromDirectory } from '../../services/import.service';
import { validateNoDuplicateContentTypeUris } from './import';
import { isEqual } from 'lodash';
import { ExportBuilderOptions } from '../../interfaces/export-builder-options.interface';
import readline from 'readline';
import DataPresenter from '../../view/data-presenter';

export const command = 'export <dir>';

export const desc = 'Export Content Types';

export const builder = (yargs: Argv): void => {
  yargs
    .positional('dir', {
      describe: 'Output directory for the exported Content Type definitions',
      type: 'string'
    })
    .option('schemaId', {
      type: 'string',
      describe: 'content-type-schema ID(s) of Content Type(s) to export',
      requiresArg: true
    })
    .array<string>('schemaId');
};

const equals = (a: ContentType, b: ContentType): boolean =>
  a.contentTypeUri === b.contentTypeUri && isEqual(a.settings, b.settings);

interface ExportRecord {
  readonly filename: string;
  readonly status: ExportResult;
  readonly contentType: ContentType;
}

export const filterContentTypesByUri = (listToFilter: ContentType[], contentTypeUriList: string[]): ContentType[] => {
  let unmatchedContentTypeUriList: string[] = [];
  let filteredList: ContentType[] = listToFilter;
  if (contentTypeUriList.length > 0) {
    filteredList = listToFilter.filter(contentType =>
      contentTypeUriList.some(uri => contentType.contentTypeUri === uri)
    );
    unmatchedContentTypeUriList = contentTypeUriList.filter(
      uri => !listToFilter.some(contentType => contentType.contentTypeUri === uri)
    );
    if (unmatchedContentTypeUriList.length > 0) {
      throw new Error(
        `The following schema ID(s) could not be found: [${unmatchedContentTypeUriList.map(u => `'${u}'`).join(', ')}].`
      );
    }
  }
  return filteredList;
};

export const getExportRecordForContentType = (
  contentType: ContentType,
  outputDir: string,
  previouslyExportedContentTypes: { [filename: string]: ContentType }
): ExportRecord => {
  const indexOfExportedContentType = Object.values(previouslyExportedContentTypes).findIndex(
    c => c.contentTypeUri === contentType.contentTypeUri
  );
  if (indexOfExportedContentType < 0) {
    return { filename: uniqueFilename(outputDir, 'json'), status: 'CREATED', contentType };
  }
  const filename = Object.keys(previouslyExportedContentTypes)[indexOfExportedContentType];
  const previouslyExportedContentType = Object.values(previouslyExportedContentTypes)[indexOfExportedContentType];
  if (equals(previouslyExportedContentType, contentType)) {
    return { filename, status: 'UP-TO-DATE', contentType };
  }
  return {
    filename,
    status: 'UPDATED',
    contentType
  };
};

type ExportsMap = {
  uri: string;
  filename: string;
};

export const getExports = (
  outputDir: string,
  previouslyExportedContentTypes: { [filename: string]: ContentType },
  contentTypesBeingExported: ContentType[]
): [ExportRecord[], ExportsMap[]] => {
  const allExports: ExportRecord[] = [];
  const updatedExportsMap: ExportsMap[] = []; // uri x filename
  for (const contentType of contentTypesBeingExported) {
    const exportRecord = getExportRecordForContentType(contentType, outputDir, previouslyExportedContentTypes);
    if (contentType.contentTypeUri) {
      allExports.push(exportRecord);
      if (exportRecord.status === 'UPDATED') {
        updatedExportsMap.push({ uri: contentType.contentTypeUri, filename: exportRecord.filename });
      }
    }
  }
  return [allExports, updatedExportsMap];
};

export const promptToOverwriteExports = (updatedExportsMap: { [key: string]: string }[]): Promise<boolean> => {
  return new Promise((resolve): void => {
    process.stdout.write('The following files will be overwritten:');
    // display updatedExportsMap as a table of uri x filename
    new DataPresenter(updatedExportsMap).render();

    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });

    rl.question('Do you want to continue (y/n)?: ', answer => {
      rl.close();
      return resolve(answer === 'y');
    });
  });
};

export const processContentTypes = async (
  outputDir: string,
  previouslyExportedContentTypes: { [filename: string]: ContentType },
  contentTypesBeingExported: ContentType[]
): Promise<void> => {
  const [allExports, updatedExportsMap] = getExports(
    outputDir,
    previouslyExportedContentTypes,
    contentTypesBeingExported
  );
  if (Object.keys(updatedExportsMap).length > 0 && !(await promptToOverwriteExports(updatedExportsMap))) {
    process.exit(1);
  }

  const tableStream = (createStream(streamTableOptions) as unknown) as TableStream;
  tableStream.write([chalk.bold('file'), chalk.bold('contentTypeUri'), chalk.bold('result')]);
  for (const { filename, status, contentType } of allExports) {
    if (status !== 'UP-TO-DATE') {
      /* eslint-disable @typescript-eslint/no-unused-vars */ // id is intentionally thrown away on the next line
      const { id, ...exportedContentType } = contentType; // do not export id
      writeJsonToFile(filename, new ContentType(exportedContentType));
    }
    tableStream.write([filename, contentType.contentTypeUri || '', status]);
  }
  process.stdout.write('\n');
};

export const handler = async (argv: Arguments<ExportBuilderOptions & ConfigurationParameters>): Promise<void> => {
  const { dir, schemaId } = argv;
  const previouslyExportedContentTypes = loadJsonFromDirectory<ContentType>(dir, ContentType);
  validateNoDuplicateContentTypeUris(previouslyExportedContentTypes);

  const client = dynamicContentClientFactory(argv);
  const hub = await client.hubs.get(argv.hubId);
  const storedContentTypes = await paginator(hub.related.contentTypes.list);
  const filteredContentTypes = filterContentTypesByUri(storedContentTypes, schemaId || []);
  await processContentTypes(dir, previouslyExportedContentTypes, filteredContentTypes);
};
