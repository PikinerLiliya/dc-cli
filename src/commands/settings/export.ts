import { Arguments, Argv } from 'yargs';
import { ConfigurationParameters } from '../configure';
import dynamicContentClientFactory from '../../services/dynamic-content-client-factory';
import paginator from '../../common/dc-management-sdk-js/paginator';
import { Hub, WorkflowState } from 'dc-management-sdk-js';
import { nothingExportedExit, promptToExportSettings, writeJsonToFile } from '../../services/export.service';
import { ExportBuilderOptions } from '../../interfaces/export-builder-options.interface';
import * as path from 'path';

export const command = 'export <dir>';

export const desc = 'Export Hub Settings';

export const builder = (yargs: Argv): void => {
  yargs.positional('dir', {
    describe: 'Output directory for the exported Settings',
    type: 'string'
  });
};

export const processSettings = async (
  outputDir: string,
  hubToExport: Hub,
  workflowStates: WorkflowState[]
): Promise<void> => {
  const {
    id,
    name,
    label,
    description,
    status,
    settings,
    createdBy,
    createdDate,
    lastModifiedBy,
    lastModifiedDate
  } = hubToExport;
  let dir = outputDir;
  if (outputDir.substr(-1) === path.sep) {
    dir = dir.slice(0, -1);
  }
  const file = path.basename(`settings-${hubToExport.id}`, '.json');

  const uniqueFilename = dir + path.sep + file + '.json';

  if (!(await promptToExportSettings(uniqueFilename))) {
    return nothingExportedExit();
  }

  writeJsonToFile(uniqueFilename, {
    id,
    name,
    label,
    description,
    status,
    settings,
    createdBy,
    createdDate,
    lastModifiedBy,
    lastModifiedDate,
    workflowStates: workflowStates
  });

  process.stdout.write('Settings exported successfully! \n');
};

export const handler = async (argv: Arguments<ExportBuilderOptions & ConfigurationParameters>): Promise<void> => {
  const { dir } = argv;

  const client = dynamicContentClientFactory(argv);
  const hub = await client.hubs.get(argv.hubId);
  const workflowStates = await paginator(hub.related.workflowStates.list);

  await processSettings(dir, hub, workflowStates);
};
