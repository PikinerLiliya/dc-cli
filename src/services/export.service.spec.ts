import fs from 'fs';
import { writeJsonToFile } from './export.service';
import { uniqueFilename } from './export.service';
import { ContentType } from 'dc-management-sdk-js';

jest.mock('fs');

describe('export service tests', () => {
  describe('uniqueFilename tests', () => {
    it('should produce unique file names', (): void => {
      const filenames: Set<string> = new Set();
      for (let n = 0; n < 100; n++) {
        const filename = uniqueFilename('my-dir', 'json');
        if (filenames.has(filename)) {
          fail(`non-unique filename ${filename}`);
        } else {
          filenames.add(filename);
        }
      }
    });
  });

  describe('writeJsonToFile tests', () => {
    it('should write the content type to the specified file', (): void => {
      const contentType = new ContentType({
        id: 'content-type-id-1',
        contentTypeUri: 'content-type-uri-1'
      });
      writeJsonToFile<ContentType>('my-filename', contentType);
      expect(fs.writeFileSync).toHaveBeenCalledWith('my-filename', JSON.stringify(contentType));
    });

    it('should throw an error if it cannot write to the file', (): void => {
      (fs.writeFileSync as jest.Mock).mockImplementationOnce(() => {
        throw new Error('Cannot write to file/directory');
      });
      expect(() => writeJsonToFile<ContentType>('my-filename', new ContentType())).toThrowError(
        /^Unable to write file: my-filename, aborting export$/
      );
    });
  });
});
