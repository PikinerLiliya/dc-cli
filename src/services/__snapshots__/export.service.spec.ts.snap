// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`export service tests nothingExportedExit should exit with an export message 1`] = `
Array [
  Array [
    "Nothing was exported, exiting.
",
  ],
]
`;

exports[`export service tests promptToOverwriteExports Should return false when the answer is "n" 1`] = `
Array [
  Array [
    "Do you want to continue (y/n)?: ",
    [Function],
  ],
]
`;

exports[`export service tests promptToOverwriteExports Should return false when the answer is "n" 2`] = `
Array [
  Array [
    "The following files will be overwritten:
",
  ],
  Array [
    "undefined
",
  ],
]
`;

exports[`export service tests promptToOverwriteExports Should return false when the answer is "n" 3`] = `
Array [
  Array [
    Array [
      Array [
        "[1muri[22m",
        "[1mfilename[22m",
      ],
      Array [
        "my-content-type-uri",
        "my-export-filename",
      ],
    ],
    Object {
      "border": undefined,
    },
  ],
  Array [
    Array [
      Array [
        "[1muri[22m",
        "[1mfilename[22m",
      ],
      Array [
        "my-content-type-uri",
        "my-export-filename",
      ],
    ],
    Object {
      "border": undefined,
    },
  ],
]
`;

exports[`export service tests promptToOverwriteExports Should return false when the answer is anything but "y" 1`] = `
Array [
  Array [
    "Do you want to continue (y/n)?: ",
    [Function],
  ],
]
`;

exports[`export service tests promptToOverwriteExports Should return false when the answer is anything but "y" 2`] = `
Array [
  Array [
    "The following files will be overwritten:
",
  ],
  Array [
    "undefined
",
  ],
]
`;

exports[`export service tests promptToOverwriteExports Should return false when the answer is anything but "y" 3`] = `
Array [
  Array [
    Array [
      Array [
        "[1muri[22m",
        "[1mfilename[22m",
      ],
      Array [
        "my-content-type-uri",
        "my-export-filename",
      ],
    ],
    Object {
      "border": undefined,
    },
  ],
  Array [
    Array [
      Array [
        "[1muri[22m",
        "[1mfilename[22m",
      ],
      Array [
        "my-content-type-uri",
        "my-export-filename",
      ],
    ],
    Object {
      "border": undefined,
    },
  ],
  Array [
    Array [
      Array [
        "[1muri[22m",
        "[1mfilename[22m",
      ],
      Array [
        "my-content-type-uri",
        "my-export-filename",
      ],
    ],
    Object {
      "border": undefined,
    },
  ],
]
`;

exports[`export service tests promptToOverwriteExports Should return true when the answer is "y" 1`] = `
Array [
  Array [
    "Do you want to continue (y/n)?: ",
    [Function],
  ],
]
`;

exports[`export service tests promptToOverwriteExports Should return true when the answer is "y" 2`] = `
Array [
  Array [
    "The following files will be overwritten:
",
  ],
  Array [
    "undefined
",
  ],
]
`;

exports[`export service tests promptToOverwriteExports Should return true when the answer is "y" 3`] = `
Array [
  Array [
    Array [
      Array [
        "[1muri[22m",
        "[1mfilename[22m",
      ],
      Array [
        "my-content-type-uri",
        "my-export-filename",
      ],
    ],
    Object {
      "border": undefined,
    },
  ],
]
`;
